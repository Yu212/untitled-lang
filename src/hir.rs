use ecow::EcoString;
use la_arena::Idx;

use crate::scope::{FnId, VarId};

type ExprIdx = Idx<Expr>;
type StmtIdx = Idx<Stmt>;

#[derive(Debug)]
pub struct Root {
    pub stmts: Vec<StmtIdx>,
}

#[derive(Debug, Clone)]
pub enum Stmt {
    LetStmt {
        var_id: Option<VarId>,
        expr: ExprIdx,
    },
    WhileStmt {
        cond: ExprIdx,
        block: ExprIdx,
    },
    BreakStmt {
    },
    ExprStmt {
        expr: ExprIdx,
    },
}

#[derive(Debug, Clone)]
pub enum Expr {
    Missing,
    Binary {
        op: BinaryOp,
        lhs: ExprIdx,
        rhs: ExprIdx,
    },
    Unary {
        op: UnaryOp,
        expr: ExprIdx,
    },
    Ref {
        var_id: Option<VarId>,
    },
    If {
        cond: ExprIdx,
        then_expr: ExprIdx,
        else_expr: Option<ExprIdx>,
    },
    FnCall {
        fn_id: Option<FnId>,
        args: Vec<ExprIdx>,
    },
    Block {
        stmts: Vec<StmtIdx>,
    },
    Literal {
        n: Option<u64>,
    },
}

#[derive(Debug)]
pub struct Identifier {
    pub name: EcoString,
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub enum BinaryOp {
    Add,
    Sub,
    Mul,
    Div,
    Rem,
    Assign,
}

impl BinaryOp {
    pub fn binding_power(&self) -> (u8, u8) {
        match self {
            BinaryOp::Add => (3, 4),
            BinaryOp::Sub => (3, 4),
            BinaryOp::Mul => (5, 6),
            BinaryOp::Div => (5, 6),
            BinaryOp::Rem => (5, 6),
            BinaryOp::Assign => (2, 1),
        }
    }
}

#[derive(Debug, Clone)]
pub enum UnaryOp {
    Neg,
}

impl UnaryOp {
    pub fn binding_power(&self) -> ((), u8) {
        match self {
            UnaryOp::Neg => ((), 7),
        }
    }
}
