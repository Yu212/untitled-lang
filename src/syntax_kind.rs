#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
#[repr(u16)]
pub enum SyntaxKind {
    Error,
    Whitespace,
    Eof,
    EqEq,
    Neq,
    Comma,
    Semicolon,
    Colon,
    Ident,
    TypedIdent,
    TypeSpec,
    IdentTypeSpec,
    ArrayTypeSpec,
    LetKw,
    IfKw,
    ElseKw,
    WhileKw,
    BreakKw,
    FunKw,
    TrueKw,
    FalseKw,
    Equals,
    Plus,
    Minus,
    Star,
    Slash,
    Percent,
    Number,
    OpenParen,
    CloseParen,
    OpenBracket,
    CloseBracket,
    OpenBrace,
    CloseBrace,
    Root,
    FuncDef,
    ParamList,
    NumberLiteral,
    BoolLiteral,
    ArrayLiteral,
    BinaryExpr,
    PrefixExpr,
    ParenExpr,
    RefExpr,
    IfExpr,
    FnCallExpr,
    BlockExpr,
    Stmt,
    LetStmt,
    WhileStmt,
    BreakStmt,
    ExprStmt,
}

impl SyntaxKind {
    pub fn is_eof(self) -> bool {
        self == Self::Eof
    }

    pub fn is_error(self) -> bool {
        self == Self::Error
    }

    pub fn is_trivia(self) -> bool {
        matches!(self, SyntaxKind::Whitespace)
    }
}

impl From<SyntaxKind> for rowan::SyntaxKind {
    fn from(kind: SyntaxKind) -> Self {
        Self(kind as u16)
    }
}
