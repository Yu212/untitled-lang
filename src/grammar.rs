use crate::hir::{BinaryOp, UnaryOp};
use crate::parser::{CompletedMarker, Parser};
use crate::syntax_kind::SyntaxKind;
use crate::token_set::TokenSet;

const RECOVERY_SET: TokenSet = TokenSet::new(&[SyntaxKind::Semicolon]);

pub fn root(p: &mut Parser<'_>) {
    let m = p.start();
    p.eat_trivia();
    while !p.at(SyntaxKind::Eof) {
        stmt(p);
    }
    m.complete(p, SyntaxKind::Root);
}

pub fn stmt(p: &mut Parser<'_>) -> CompletedMarker {
    match p.current() {
        SyntaxKind::LetKw => let_stmt(p),
        SyntaxKind::WhileKw => while_stmt(p),
        SyntaxKind::BreakKw => break_stmt(p),
        _ => expr_stmt(p),
    }
}

pub fn let_stmt(p: &mut Parser<'_>) -> CompletedMarker {
    assert!(p.at(SyntaxKind::LetKw));
    let m = p.start();
    p.bump();
    p.expect(SyntaxKind::Ident);
    p.expect(SyntaxKind::Equals);
    expr(p, 0);
    p.expect(SyntaxKind::Semicolon);
    m.complete(p, SyntaxKind::LetStmt)
}

pub fn while_stmt(p: &mut Parser<'_>) -> CompletedMarker {
    assert!(p.at(SyntaxKind::WhileKw));
    let m = p.start();
    p.bump();
    expr(p, 0);
    block_expr(p);
    m.complete(p, SyntaxKind::WhileStmt)
}

pub fn break_stmt(p: &mut Parser<'_>) -> CompletedMarker {
    assert!(p.at(SyntaxKind::BreakKw));
    let m = p.start();
    p.bump();
    p.expect(SyntaxKind::Semicolon);
    m.complete(p, SyntaxKind::BreakStmt)
}

pub fn expr_stmt(p: &mut Parser<'_>) -> CompletedMarker {
    let m = p.start();
    expr(p, 0);
    p.expect(SyntaxKind::Semicolon);
    m.complete(p, SyntaxKind::ExprStmt)
}

pub fn expr(p: &mut Parser<'_>, min_binding_power: u8) -> Option<CompletedMarker> {
    let mut lhs = lhs(p)?;
    loop {
        let op = match p.current() {
            SyntaxKind::Plus => BinaryOp::Add,
            SyntaxKind::Minus => BinaryOp::Sub,
            SyntaxKind::Star => BinaryOp::Mul,
            SyntaxKind::Slash => BinaryOp::Div,
            SyntaxKind::Percent => BinaryOp::Rem,
            SyntaxKind::Equals => BinaryOp::Assign,
            _ => break,
        };
        let (left_binding_power, right_binding_power) = op.binding_power();
        if left_binding_power < min_binding_power {
            break;
        }
        p.bump();
        let m = lhs.precede(p);
        let rhs = expr(p, right_binding_power);
        lhs = m.complete(p, SyntaxKind::BinaryExpr);
        if rhs.is_none() {
            break;
        }
    }
    Some(lhs)
}

pub fn lhs(p: &mut Parser<'_>) -> Option<CompletedMarker> {
    match p.current() {
        SyntaxKind::Number => Some(number(p)),
        SyntaxKind::Minus => Some(prefix_expr(p)),
        SyntaxKind::OpenParen => Some(paren_expr(p)),
        SyntaxKind::OpenBrace => Some(block_expr(p)),
        SyntaxKind::Ident if p.nth_at(1, SyntaxKind::OpenParen) => Some(fn_call_expr(p)),
        SyntaxKind::Ident => Some(ref_expr(p)),
        SyntaxKind::IfKw => Some(if_expr(p)),
        _ => {
            p.error_and_recover(&[SyntaxKind::Number, SyntaxKind::Minus, SyntaxKind::OpenParen, SyntaxKind::OpenBrace, SyntaxKind::Ident], &RECOVERY_SET);
            None
        }
    }
}

pub fn prefix_expr(p: &mut Parser<'_>) -> CompletedMarker {
    assert!(p.at(SyntaxKind::Minus));
    let m = p.start();
    p.bump();
    let op = UnaryOp::Neg;
    let ((), right_binding_power) = op.binding_power();
    expr(p, right_binding_power);
    m.complete(p, SyntaxKind::PrefixExpr)
}

pub fn paren_expr(p: &mut Parser<'_>) -> CompletedMarker {
    assert!(p.at(SyntaxKind::OpenParen));
    let m = p.start();
    p.bump();
    expr(p, 0);
    p.expect(SyntaxKind::CloseParen);
    m.complete(p, SyntaxKind::ParenExpr)
}

pub fn ref_expr(p: &mut Parser<'_>) -> CompletedMarker {
    assert!(p.at(SyntaxKind::Ident));
    let m = p.start();
    p.bump();
    m.complete(p, SyntaxKind::RefExpr)
}

pub fn if_expr(p: &mut Parser<'_>) -> CompletedMarker {
    assert!(p.at(SyntaxKind::IfKw));
    let m = p.start();
    p.bump();
    expr(p, 0);
    block_expr(p);
    if p.eat(SyntaxKind::ElseKw) {
        block_expr(p);
    }
    m.complete(p, SyntaxKind::IfExpr)
}

pub fn fn_call_expr(p: &mut Parser<'_>) -> CompletedMarker {
    assert!(p.at(SyntaxKind::Ident));
    let m = p.start();
    p.bump();
    p.expect(SyntaxKind::OpenParen);
    if !p.at(SyntaxKind::CloseParen) {
        expr(p, 0);
        while p.eat(SyntaxKind::Comma) {
            expr(p, 0);
        }
    }
    p.expect(SyntaxKind::CloseParen);
    m.complete(p, SyntaxKind::FnCallExpr)
}

pub fn block_expr(p: &mut Parser<'_>) -> CompletedMarker {
    let m = p.start();
    p.expect(SyntaxKind::OpenBrace);
    while !p.at(SyntaxKind::CloseBrace) && !p.at(SyntaxKind::Eof) {
        stmt(p);
    }
    p.expect(SyntaxKind::CloseBrace);
    m.complete(p, SyntaxKind::BlockExpr)
}

pub fn number(p: &mut Parser<'_>) -> CompletedMarker {
    assert!(p.at(SyntaxKind::Number));
    let m = p.start();
    p.bump();
    m.complete(p, SyntaxKind::Literal)
}

#[cfg(test)]
mod tests {
    use crate::diagnostic::Diagnostic;
    use crate::language::SyntaxNode;
    use crate::lexer::Lexer;
    use crate::parser::Parser;

    fn parse(text: &str) -> (SyntaxNode, Vec<Diagnostic>) {
        let lexer = Lexer::new(text);
        let (tokens, _) = lexer.tokenize();
        let parser = Parser::new(tokens);
        parser.parse()
    }

    #[test]
    fn multi_stmt() {
        insta::assert_debug_snapshot!(parse("1; 2; 3;"));
    }

    #[test]
    fn let_stmt() {
        insta::assert_debug_snapshot!(parse("let a = 1;"));
    }

    #[test]
    fn expr_stmt() {
        insta::assert_debug_snapshot!(parse("1 + 2 * 3;"));
    }

    #[test]
    fn unary() {
        insta::assert_debug_snapshot!(parse("-1 * -2;"));
    }

    #[test]
    fn paren() {
        insta::assert_debug_snapshot!(parse("(1 + 2) * 3;"));
    }

    #[test]
    fn ref_expr() {
        insta::assert_debug_snapshot!(parse("a + bc;"));
    }

    #[test]
    fn fn_call_expr() {
        insta::assert_debug_snapshot!(parse("f(a, 1);"));
    }

    #[test]
    fn block_expr() {
        insta::assert_debug_snapshot!(parse("{1;} + {2;};"));
    }
}
