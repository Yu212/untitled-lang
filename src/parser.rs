use drop_bomb::DropBomb;
use rowan::TextRange;

use crate::event;
use crate::event::Event;
use crate::grammar::root;
use crate::language::SyntaxNode;
use crate::syntax_error::SyntaxError;
use crate::syntax_kind::SyntaxKind;
use crate::token::Token;

pub struct Parser<'a> {
    tokens: Vec<Token<'a>>,
    cursor: usize,
    events: Vec<Event>,
}

impl<'a> Parser<'a> {
    pub fn new(tokens: Vec<Token<'a>>) -> Self {
        Self {
            tokens,
            cursor: 0,
            events: Vec::new(),
        }
    }
    pub fn parse(mut self) -> (SyntaxNode, Vec<SyntaxError>) {
        root(&mut self);
        event::process(&mut self.events, self.tokens)
    }
    pub fn current(&self) -> SyntaxKind {
        self.nth(0)
    }
    pub fn nth(&self, n: usize) -> SyntaxKind {
        let mut temp_cursor = self.cursor;
        for _ in 0..n {
            if self.tokens[temp_cursor].kind.is_eof() {
                break;
            }
            temp_cursor += 1;
            while self.tokens[temp_cursor].kind.is_trivia() {
                temp_cursor += 1;
            }
        }
        self.tokens[temp_cursor].kind.clone()
    }
    fn current_range(&self) -> TextRange {
        self.tokens[self.cursor].range
    }
    pub fn at(&self, kind: SyntaxKind) -> bool {
        self.nth_at(0, kind)
    }
    pub fn at_set(&self, set: &[SyntaxKind]) -> bool {
        set.contains(&self.current())
    }
    pub fn nth_at(&self, n: usize, kind: SyntaxKind) -> bool {
        self.nth(n) == kind
    }
    pub fn eat(&mut self, kind: SyntaxKind) -> bool {
        if self.at(kind) {
            self.bump();
            true
        } else {
            false
        }
    }
    pub fn assert(&mut self, kind: SyntaxKind) {
        assert!(self.at(kind));
        self.bump();
    }
    pub fn bump(&mut self) {
        if self.current() != SyntaxKind::Eof {
            self.consume_token();
            self.eat_trivia();
        }
    }
    pub fn expect(&mut self, kind: SyntaxKind) -> bool {
        if self.eat(kind) {
            true
        } else {
            self.error(format!("expected: {kind:?}"));
            false
        }
    }
    pub fn eof(&self) -> bool {
        self.at(SyntaxKind::Eof)
    }
    pub fn error(&mut self, message: impl Into<String>) {
        self.events.push(Event::Error(SyntaxError {
            message: message.into(),
            range: self.current_range(),
        }));
    }
    pub fn error_and_bump(&mut self, message: impl Into<String>) {
        let m = self.start();
        self.error(message);
        self.bump();
        m.complete(self, SyntaxKind::Error);
    }
    pub fn error_and_recover(&mut self, message: impl Into<String>, recovery: &[SyntaxKind]) {
        if self.at_set(recovery) {
            self.error(message);
            return;
        }
        self.error_and_bump(message);
    }
    pub fn eat_trivia(&mut self) {
        while self.current().is_trivia() {
            self.consume_token();
        }
    }
    pub fn start(&mut self) -> Marker {
        let pos = self.events.len();
        self.events.push(Event::Placeholder);
        Marker {
            pos,
            bomb: DropBomb::new("Marker must be either completed or abandoned")
        }
    }
    fn consume_token(&mut self) {
        self.events.push(Event::Token(self.current()));
        self.cursor += 1;
    }
}

pub struct Marker {
    pos: usize,
    bomb: DropBomb,
}

impl Marker {
    pub fn complete(mut self, p: &mut Parser<'_>, kind: SyntaxKind) -> CompletedMarker {
        self.bomb.defuse();
        let event = &mut p.events[self.pos];
        *event = Event::StartNode(kind, None);
        p.events.push(Event::FinishNode);
        CompletedMarker {
            pos: self.pos
        }
    }
    pub fn abandon(mut self) {
        self.bomb.defuse();
    }
}

pub struct CompletedMarker {
    pos: usize,
}

impl CompletedMarker {
    pub fn precede(self, p: &mut Parser<'_>) -> Marker {
        let m = p.start();
        match p.events[self.pos] {
            Event::StartNode(_, ref mut forward_parent) => {
                *forward_parent = Some(m.pos);
            },
            _ => unreachable!(),
        }
        m
    }
}
