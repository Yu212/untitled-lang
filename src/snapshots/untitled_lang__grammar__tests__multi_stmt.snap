---
source: src/grammar.rs
expression: "parse(\"1; 2; 3;\")"
---
(
    Root@0..8
      ExprStmt@0..3
        Literal@0..1
          Number@0..1 "1"
        Semicolon@1..2 ";"
        Whitespace@2..3 " "
      ExprStmt@3..6
        Literal@3..4
          Number@3..4 "2"
        Semicolon@4..5 ";"
        Whitespace@5..6 " "
      ExprStmt@6..8
        Literal@6..7
          Number@6..7 "3"
        Semicolon@7..8 ";"
    ,
    [],
)
